add test selector and code refactoring

squirrelsc opened the pull request:
1. implement test selector 
1. add unittest
1. refactoring code to use collection for platforms, environments, nodes, and tools.
1. add statistics and test result for test case progress
1. refactoring code to be pythonic, including replacing singleton to global objects, merge files and packages, remove factory patterns.
1. enable schema validation.
1. inherit from unittest to get all assert methods
1. add perf_timer for profiling, and improve current profiling
1. refin log to output meaningful prefix
1. split session to local and ssh sessions.
1. add lightweight tool wrapper to use tool easier
1. add document for tool.py
1. add __init__.py for mypy checking
1. other minor improvements

squirrelsc committed:
add statistics and test result for test progress

1. add statistics and test result for test case progress
2. inherit from unittest to get all assert methods
3. add perf_timer for profiling, and improve current profiling
4. add document for tool.py
5. add init.py for mypy checking
6. other minor improvements

squirrelsc committed:
provide a simple way to use tool

1. add LightTool to support oneline call to script
2. add default function for tool class

squirrelsc committed:
refine log to output meaningful prefix

squirrelsc committed:
replace singleton to global object

1. global object is more pythonic implementation, so change it.
2. Config is not used widely, so remove singleton.

squirrelsc committed:
merge files to be pythonic

Python file is a module, which is not used to put only one thing.
All related things should be put in same module.

platform.py causes error in uuid.py, so change name to platform_.py

squirrelsc committed:
remove factory patterns from code

Use pythonic function to replace factory.

squirrelsc committed:
split code of local and ssh session

squirrelsc committed:
Improve logger to use customized Logger class.

squirrelsc committed:
Add schema validation

1. add cerberus to validate schema.
2. remove full.yml, and add schema.yml as schema
3. fix schema due to new schema

squirrelsc committed:
Implement test case selector

1. add testselector to deal with criteria and apply settings
2. improve logging to support ut.
3. remove TestSuiteData class, as it doesn't need.
4. add ut in github action.
5. other minor improvements.

andyleejordan reviewed (approved):
Looking good. I appreciate a lot of the refactors. Left a couple suggestions and spotted one bug (maybe fixed since I added the comment). There are a few files I need to spend a lot more time digging into, but won't hold this up to do that.

squirrelsc committed:
refactor: use dict for collections

1. add environments, nodes, platforms classes like a collection.
2. add tools to use a tool like [ToolType], and move related code to
    executable.
3. other minor improvements.
